packages/programs/src/resolve-transaction-error.ts

```diff
@@ -0,0 +1,53 @@
+import type { Address } from '@solana/addresses';
+import type { Transaction } from '@solana/transactions';
+
+import { Program, ProgramWithErrors } from './program';
+
+/**
+ * Resolves a custom program error from a transaction error
+ * with logs using the provided list of programs.
+ * The original error is returned if the error cannot be
+ * resolved from the given programs.
+ *
+ * @param error The raw error to resolve containing the program logs.
+ * @param transaction The transaction that caused the error.
+ * @param programs The list of programs to go through when resolving the transaction error.
+ * They should ideally contain all programs the transaction is sending instructions to.
+ * @returns The resolved program error, or the original transaction error
+ * if the error cannot be resolved using the provided programs.
+ */
+export function resolveTransactionError(
+    error: Error & { logs?: string[] },
+    transaction: Transaction,
+    programs: Program[],
+): Error {
+    // Compute the full logs from which to parse the instruction index and error code.
+    const fullLogs = error.message + '\n' + (error.logs ?? []).join('\n');
```

Comment:
I guess the thinking here is that building a string and then running one `match` over it is faster than building a state machine?

```ts
const enum State {
  SEARCHING_FOR_INDEX,
  SEARCHING_FOR_CUSTOM_ERROR,
};

let state;
for (const log in logs) {
  switch (state.__type) {
    case State.SEARCHING_FOR_INDEX:
    {
      const index = log.match(/Error processing Instruction (\d+)/)?.[1];
      if (index != null) {
        state = {__type: State.SEARCHING_FOR_CUSTOM_ERROR, instructionIndex: index};
      }
      break;
    }
    case State.SEARCHING_FOR_CUSTOM_ERROR:
    {
      /* ... */
    }
  }
}
```

Reply:
You need to find both of them in order to be able to resolve the error so it's not the case of switching between different strategies. There are situations were the index and the error code may sometimes be present in the message and not in the logs, and vice versa. By combining the whole thing together and searching that, we offer a more robust solution.

Reply:
That's not a strategy switch, it's a state machine. Iterate over logs until you find the instruction by index, transition to the next state where you search for the custom error, etc…

Reply:
Gotcha. I prefer leaving this as-is because IMO the invariant "the instruction index needs to appear before the error code" isn't necessary.